### PR TITLE
Serialize and deserialize null

### PR DESCRIPTION
**Pull request by rsgowman**
Build nanopb generated protos and use them to [de]serialize null values. (Exciting, I know.)

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 1]
```diff
@@ -16,16 +16,86 @@
...
+
+Serializer::ValueWithType Serializer::DecodeValueWithType(const uint8_t* bytes,
+                                                          size_t length) {
+  pb_istream_t stream = pb_istream_from_buffer(bytes, length);
+  pb_wire_type_t wire_type;
+  uint32_t tag;
+  bool eof;
+  bool status = pb_decode_tag(&stream, &wire_type, &tag, &eof);
+  if (!status || wire_type != PB_WT_VARINT ||
+      tag != google_firestore_v1beta1_Value_null_value_tag) {
```
Comment (zxu123) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Might follow the same code structure as in other APIs of Serializer, where there is a switch with only null type dealt rather than combining with the check of status, which is type-independant.

File: Firestore/core/src/firebase/firestore/remote/serializer.h [hunk 2]
```diff
@@ -17,27 +17,109 @@
...
-  Serializer(/*DatabaseID databaseId*/);
+  /**
+   * @brief Wraps (nanopb) google_firestore_v1beta1_Value with type information.
+   */
+  struct ValueWithType {
+    firebase::firestore::model::FieldValue::Type type;
+    google_firestore_v1beta1_Value value;
+  };
+
+  explicit Serializer(firebase::firestore::model::DatabaseId database_id)
```
Comment (zxu123) [inline Firestore/core/src/firebase/firestore/remote/serializer.h]: Should accept pointer (and not owning). DatabaseIs is shared.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc [hunk 3]
```diff
@@ -17,12 +17,86 @@
...
+
+  void ExpectRoundTrip(const Serializer::ValueWithType& proto,
+                       const uint8_t* bytes,
+                       size_t bytes_len,
+                       FieldValue::Type type) {
+    EXPECT_EQ(type, proto.type);
+    // TODO(rsgowman): How big should this buffer be? Unclear; see TODO in
+    // remote/serializer.h on the Serializer::EncodeValueWithType() method.
+    // Hardcode to 1k for now.
+    uint8_t actual_bytes[1024];
```
Comment (zxu123) [inline Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: Can you also add a check to make sure bytes_len < 1024 to avoid overflow the buffer early (and potentially catch that explicitly, if that do happen).

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 4]
```diff
@@ -16,16 +16,86 @@
 
 #include "Firestore/core/src/firebase/firestore/remote/serializer.h"
 
-// TODO(rsgowman): These are (currently!) unnecessary includes. Adding for now
-// to ensure we can find nanopb's generated header files.
-#include "Firestore/Protos/nanopb/google/firestore/v1beta1/document.pb.h"
-#include "Firestore/Protos/nanopb/google/protobuf/timestamp.pb.h"
+#include <pb_decode.h>
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: C++ doesn't much care.

I think we use <> for system includes and "" for local includes. I think external libraries fall into the 'system include' bucket. (eg we use '#include <openssl/whatever.h>'). Therefore, I've left as is.

Exception: Our use of absl is inconsistent. We use both <> and ""..  :(

File: Firestore/core/src/firebase/firestore/remote/serializer.h [hunk 5]
```diff
@@ -17,27 +17,109 @@
...
+  switch (lhs.type) {
+    case firebase::firestore::model::FieldValue::Type::Null:
+      FIREBASE_DEV_ASSERT(lhs.value.null_value ==
+                          google_protobuf_NullValue_NULL_VALUE);
+      FIREBASE_DEV_ASSERT(rhs.value.null_value ==
+                          google_protobuf_NullValue_NULL_VALUE);
+      return true;
+    default:
+      // TODO(rsgowman): implement the other types
+      abort();
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: We seem to be using <X.h> rather than <cX>, so I've included stdlib.h. (It's my expectation that these abort()s will be eliminated shortly in any case.)

File: Firestore/core/src/firebase/firestore/remote/serializer.h [hunk 6]
```diff
@@ -17,27 +17,109 @@
...
+   *
+   * @return The (nanopb) proto equivalent of the bytes.
+   */
+  // TODO(rsgowman): error handling.
+  // TODO(rsgowman): do we want to indicate how many bytes were actually used?
+  // (Probably; if so, we'll add that when we need it.)
+  static ValueWithType DecodeValueWithType(const uint8_t* bytes, size_t length);
+
+ private:
+  firebase::firestore::model::DatabaseId database_id_;
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: Currently, it isn't. The other platforms use it in some of the (not yet present) methods of the serialization class. It's perhaps premature for me to declare the ctor and instance var... shrug. I've left it in for now.

File: Firestore/core/src/firebase/firestore/remote/serializer.h [hunk 7]
```diff
@@ -17,27 +17,109 @@
...
+   * buffer. On output, will contain the number of bytes actually used.
+   */
+  // TODO(rsgowman): error handling, incl return code.
+  // TODO(rsgowman): how large should the output buffer be? The docs on this
+  // method give no indication. (Unfortunately, neither does the underlying
+  // nanopb calls; we may be stuck with something vague.) But we can do some
+  // further investigation to see if there's some way to get nanopb to tell us
+  // how much space it's going to need and then expose/document that here.
+  static void EncodeValueWithType(const ValueWithType& value,
+                                  uint8_t* out_bytes,
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: Done.

[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: > Very optional: perhaps absl::Span could be used here. It's a string_view-like non-owning wrapper that holds pointer to array and the size together. Then this function could return the number of bytes actually written and avoid an input/output parameter (though if you plan to return an error code, then this doesn't apply).

I'm planning to add a return code (as soon as I get around to handling errors.) That said, I don't object to using Span anyway, though we don't currently have it in our code base, so I'm reluctant to add it just for this.

> Closely related, I'm kinda wary about the input/output parameter. Does the caller always care about the number written? 

I would think so. It'll be the code that interacts with grpc, and while I'm not sure how that works yet, at some point, it'll need to know how many bytes to write to the network.

> If not, then maybe something like (const ValueWithType& value, uint8_t* out_bytes, size_t size, size_t* bytes_written = nullptr)?

> Also, it makes sense for Nanopb to leave the allocation of the buffer to the caller. Does the same apply to the Serializer, or perhaps it could take care of the allocation strategy as well (assuming there is allocation involved. Maybe the callers use static buffers)? If that were the case, then maybe the signature could hide the low-level details of communication with Nanopb and accept/return some easier-to-use type (say, vector<char>/unique_ptr<char[]>)?

Again, I'm not sure just yet, but my assumption is that something that interfaces with grpc will be the caller here. It wouldn't be unreasonable for grpc to simply have a static buffer that it reuses... but I simply don't know enough about how that works just yet.

We've already got a TODO there related to allocation, so I'll add your suggestion to it, but will otherwise leave this as is for now until we have a better idea on how it'll be used.

File: Firestore/core/src/firebase/firestore/remote/serializer.h (hunk 2 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: Copying by value was deliberate... but also lazy on my part. (Assumptions: (a) DatabaseId contains two fairly small strings and is thus cheap to copy; (b) Only a small number of Serializer objects will be constructed, typically 1, and thus the DAtabaseId won't be copied often.)

I was being lazy, because I don't know the lifetime of DatabaseIds. In other words, can we assume DatabaseIds are long-lived? (i.e. a const& would be fine as opposed to a shared_ptr?)

File: Firestore/core/src/firebase/firestore/remote/serializer.h [hunk 8]
```diff
@@ -17,27 +17,109 @@
...
 // protocol buffer is missing a critical field or has a value we can't
 // interpret." Adjust for C++.
 class Serializer {
  public:
-  // TODO(rsgowman): Adjust ctor to accept a DatabaseID... once that exists.
-  Serializer(/*DatabaseID databaseId*/);
+  /**
+   * @brief Wraps (nanopb) google_firestore_v1beta1_Value with type information.
+   */
+  struct ValueWithType {
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: I like that better. Done.

**Comment by rsgowman** [comment]
@var-const Re: PR desc; yeah, I'll expand on that a bit. Alternatively, we can split this into multiple PRs.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: C++ doesn't care but we should because we have to write double the copybara rules to cover either pattern :-(.

The general rule is:
  * angle brackets for system headers (C and C++)
  * angle brackets for Objective-C frameworks
  * quotes for everything else

The quotes for everything else is largely a google3-ism but our release builds are coming out of there for the foreseeable future so there's no reason to deviate.

File: Firestore/core/src/firebase/firestore/remote/serializer.h (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: Just use a non-owning raw pointer.

FirestoreClient owns the lifetime of the entire constellation of objects here including DatabaseIds and Serializers.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 9]
```diff
@@ -16,16 +16,86 @@
...
 namespace firebase {
 namespace firestore {
 namespace remote {
 
-Serializer::Serializer() {
+using firebase::firestore::model::FieldValue;
+
+Serializer::ValueWithType Serializer::EncodeFieldValue(
+    const FieldValue& field_value) {
+  Serializer::ValueWithType proto_value{
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Ah, sorry, I didn't realizer that the `_init_default` is necessary for correct initialization, I assumed it was just a placeholder. I presume you need TypedValue to be POD?

File: Firestore/core/src/firebase/firestore/remote/serializer.h [hunk 10]
```diff
@@ -17,27 +17,109 @@
...
+      : database_id_(database_id) {
+  }
+
+  /**
+   * Converts the FieldValue model passed into the Value proto equivalent.
+   *
+   * @param field_value the model to convert.
+   * @return the proto representation of the model.
+   */
+  static Serializer::ValueWithType EncodeFieldValue(
```
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: Friendly ping on this question.

Now that I think about it, if `TypedValue` is the only thing that's going to be converted like that, perhaps the `Encode`/`Decode` methods should be moved into it?

File: Firestore/core/src/firebase/firestore/remote/serializer.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: Note that Travis build is failing because of the warning/error about an unused member variable.

File: Firestore/core/src/firebase/firestore/remote/serializer.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: I'm not a huge fan of raw buffers everywhere. I'm particularly not a fan of needing to precompute the size of the buffer or fail if not large enough. Better to make the buffer dynamic and allow each method to grow the buffer as needed.

To answer some of the other questions: gRPC does not assume a static buffer nor does it particularly go out of its way to make that possible. In typically usage with protobuf the buffer is malloced for each RPC.

Ultimately in the C core, the buffer is attached to the `send_message` of the `grpc_op`, which is here:
https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/grpc_types.h#L528

gRPC expects to own (and destroy) that buffer. The Objective-C wrapper serializes into an `NSData` and copies that into a buffer it can hand off into gRPC:
https://github.com/grpc/grpc/blob/86accb1768cc860a7579bb5fff9bd3c2ba7274af/src/objective-c/GRPCClient/private/NSData%2BGRPC.m#L82

I'd suggest simplifying and just doing a vector and appending as we go. If profiling turns up that we're spending a lot of time copying buffers then we can undertake the work of using slices and buffers optimally. However, I'll be _shocked_ if `memcpy` ever becomes out bottleneck.

File: Firestore/core/test/firebase/firestore/remote/serializer_test.cc (hunk 3 above)
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/remote/serializer_test.cc]: This is exactly the kind of thing I think we should eliminate by accumulating our results in a dynamic buffer. There's no reason to take on the complexity required to maintain these sizes across callers.

**Comment by wilhuff** [comment]
For future reference I'd consider b27d6a8 to be a good point to split this PR.

File: Firestore/core/src/firebase/firestore/remote/serializer.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.h]: Alright; lets use a vector for now. (Note that we still have the same problem of precomputing the size of the buffer when dealing with nanopb... but that problem is now buried in the implementation details of this class.)

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 11]
```diff
@@ -16,16 +16,94 @@
...
+                                  std::vector<uint8_t>* out_bytes) {
+  bool status;
+  // TODO(rsgowman): how large should the output buffer be? Do some
+  // investigation to see if we can get nanopb to tell us how much space it's
+  // going to need.
+  uint8_t buf[1024];
+  pb_ostream_t stream = pb_ostream_from_buffer(buf, sizeof(buf));
+  switch (value.type) {
+    case FieldValue::Type::Null:
+      status = pb_encode_tag(&stream, PB_WT_VARINT,
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Since nanopb requires us to get down to the level of encoding the tags for ourselves I'd like for us to build up a library of ways to concisely be able to describe this. Consider as an example, the code to compose leveldb keys (which operate at mostly the same granularity):

https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/Local/FSTLevelDBKey.mm#L456

Since we have error handling to consider that isn't a consideration there, perhaps we should embed the buffer and status in a writer-type object (say e.g. `ProtoWriter`) to avoid littering this code as much with error checking.

Total straw man:
```c++
class ProtoWriter {
 public:
  EncodeTag(uint32_t tag) {
    if (!last_status) return;

    last_status = pb_encode_tag(&stream_, PB_WT_VARINT, tag);
  }

  void EncodeVarInt(int64_t value) {
    if (!last_status_) return;
    last_status_ = pb_encode_varint(&stream_, value);
  }

  // maybe
  EncodeTaggedVarInt(uint32_t tag, int64_t value) {
    EncodeTag(tag);
    EncodeVarInt(value);
  }

  util::Status AppendTo(std::vector<uint8_t>* out_bytes, const char* description) {
    if (!last_status) {
      // compose error status using description
      // abort() for now for lack of Status.
    }
    // other checking
    out_bytes->insert(out_bytes->end(), buf_, buf_ + stream_.bytes_written);
    return util::OkStatus();
  }

 private:
  bool last_status_;
  uint8_t buf_[1024];
  pb_ostream_t stream_;
};
```
And this code becomes:
```c++
ProtoWriter writer;
switch (value.type) {
  case FieldValue::Type::Null:
    writer.EncodeTag(google_firestore_v1beta1_Value_null_value_tag);
    writer.EncodeVarInt(value.value.null_value);
    return writer.Append(out_bytes, "Null value");
    break;
  }
}
```
In general I want us to abstract away the fiddly details of the proto wire encoding, weird buffering, and the raw nature of the nanopb API so that we can still understand and change this code within a few months.

With that said, if these patterns aren't yet emerging it may not yet be time for this kind of abstraction.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 12]
```diff
@@ -16,8 +16,8 @@
 
 #include "Firestore/core/src/firebase/firestore/remote/serializer.h"
 
-#include <pb_decode.h>
-#include <pb_encode.h>
+#include "external/nanopb/src/nanopb/pb_decode.h"
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: On xcode, #include <nanopb/pb*.h> works (as does #include <pb*.h>)

I'm not aware of any package builds either (though I didn't look beyond my work and personal desktops which are both apt based) but a make install would put pb*.h into /usr/include (or possibly /usr/local/include or more generally $prefix/include). All of nanopbs docs use #include <pb*.h>. (And rather notably, the nanopb *generated* code uses #include <pb*.h>.) Of course, there's nothing stopping debian/etc to put the headers into /usr/include/nanopb_or_anywhere_else_really anyways, and setup the include paths via pkg-config (or to not do that and include adjusted samples as part of the .deb.)

I think FindNanopb.cmake could be adjusted to alter the paths such that #include <nanopb/pb*.h> works. But then the generated code would break (unless we added some sort of fixup pass.)

tl;dr: I'm switching back to #include <pb*.h>. Alternative: use <nanopb/pb*.h> and run a fixup pass on the generated code.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 11 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Agree with all of this, but would like to defer it. (In fact, the next PR already has some of this implemented.)

Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: SGTM. This took shape somewhat organically in the leveldb case too until I finally broke down and did the grand refactoring. As long as this is top of mind we're good for now.
